fold/seqfold.go

```diff
@@ -0,0 +1,173 @@
+package fold
+
+import (
+	"fmt"
+	"math"
+	"strings"
+
+	"github.com/TimothyStiles/poly/checks"
+)
+
+// MultibranchEnergies holds the a, b, c, d in a linear multi-branch energy
+// change function.
+// Inferred from:
+// Supplemental Material: Annu.Rev.Biophs.Biomol.Struct.33:415-40
+// doi: 10.1146/annurev.biophys.32.110601.141800
+// The Thermodynamics of DNA Structural Motifs, SantaLucia and Hicks, 2004
+type MultibranchEnergies struct {
+	A, B, C, D float64
+}
+
+// Energy holds two energies, enthaply and entropy
+// SantaLucia & Hicks (2004), Annu. Rev. Biophys. Biomol. Struct 33: 415-440
+type Energy struct {
+	// enthalpy
+	EnthalpyH float64
+	// entropy
+	EntropyS float64
+}
+
+// BPEnergy is the energy of matching base pairs
+type BPEnergy map[string]Energy
+
+// LoopEnergy is a map[int]Energy where the int is the length of the loop
+type LoopEnergy map[int]Energy
+
+// Energies holds the needed energy maps, BpEnergy and LoopEnergy, to compute
+// the folding, it also holds the complement map for the kind of sequence, RNA
+// or DNA
+type Energies struct {
+	BulgeLoops    LoopEnergy
+	Complement    map[byte]byte
+	DE            BPEnergy
+	HairpinLoops  LoopEnergy
+	Multibranch   MultibranchEnergies
+	InternalLoops LoopEnergy
+	INTERNAL_MM   BPEnergy
+	NN            BPEnergy
+	TERMINAL_MM   BPEnergy
+	TriTetraLoops BPEnergy
+}
+
+// Subsequence represent an interval of bases in the sequence that can contain
+// a inward structure.
+type Subsequence struct {
+	Start, End int
+}
+
+// A single structure with a free energy, description, and inward children
+type NucleicAcidStructure struct {
+	// Desc is the description of the NucleicAcidStructure
+	Desc string
+	// Inner is list of inner structure represented as intervals in the
+	// sequence
+	Inner []Subsequence
+	// Energy is the energy of the NucleicAcidStructure
+	Energy float64
+}
+
+func (structure NucleicAcidStructure) Equal(other NucleicAcidStructure) bool {
+	if len(structure.Inner) != len(other.Inner) {
+		return false
+	}
+	for i, val := range structure.Inner {
+		if val != other.Inner[i] {
+			return false
+		}
+	}
+	return structure.Energy == other.Energy
+}
+
+func (structure NucleicAcidStructure) Valid() bool {
+	return structure.Energy != math.Inf(1) && structure.Energy != math.Inf(-1)
+}
+
+func (structure NucleicAcidStructure) String() string {
+	i, j := "", ""
+	if len(structure.Inner) > 0 {
+		i, j = fmt.Sprint(structure.Inner[0].Start), fmt.Sprint(structure.Inner[0].End)
+	}
+	return fmt.Sprintf("%4s %4s % 6.2f  %-15s", i, j, structure.Energy, structure.Desc)
+}
+
+// MultiString returns all the fields as strings, this is useful to output the
+// result in a tabular fashion using the same format string.
+func (structure NucleicAcidStructure) MultiString() (string, string, string, string) {
```

Comment:
tests don't reach. Dead code?

Reply:
Handy method to replicate the same output as the python code, I can add a test but I'm not entirely sure if this code belongs here or not.

Reply:
should probably just remove it and `string`